hazelcast/src/main/java/com/hazelcast/query/impl/getters/Extractors.java

```diff
@@ -52,45 +57,80 @@ public Extractors(List<MapAttributeConfig> mapAttributeConfigs) {
         this.argumentsParser = new DefaultArgumentParser();
     }
 
-    public Object extract(Object targetObject, String attributeName) {
+    public Object extract(InternalSerializationService serializationService, Object target, String attributeName) {
+        Object targetObject = getTargetObject(serializationService, target);
         if (targetObject != null) {
-            Getter getter = getGetter(targetObject, attributeName);
+            Getter getter = getGetter(serializationService, targetObject, attributeName);
             try {
-                return getter.getValue(targetObject);
+                return getter.getValue(targetObject, attributeName);
             } catch (Exception ex) {
-                if (ex instanceof RuntimeException) {
-                    throw (RuntimeException) ex;
-                }
                 throw new QueryException(ex);
             }
         }
         return null;
     }
 
-    Getter getGetter(Object targetObject, String attributeName) {
+    /**
+     * @return Data (in this case it's portable) or Object (in this case it's non-portable)
+     */
+    private static Object getTargetObject(InternalSerializationService serializationService, Object target) {
+        Data targetData;
+        if (target instanceof Portable) {
+            targetData = serializationService.toData(target);
+            if (targetData.isPortable()) {
+                return targetData;
+            }
+        }
+
+        if (target instanceof Data) {
+            targetData = (Data) target;
+            if (targetData.isPortable()) {
+                return targetData;
+            }
+        }
+
+        // convert non-portable Data to object
+        if (target instanceof Data) {
```

Comment:
this check might be redundant as serialization service is doing the same thing. however it adds to readability?